### DISCUSSION TITLE
Issue refetching data

### DISCUSSION THREAD

**Discussion by kcsujeet** [TanStack/query#6025] [Q&A] [answered]
I am using `useQuery` hook to fetch a list of users as: 

```
  const { data, isLoading, isError, refetch, fetchStatus } = useQuery({
    queryKey: ['users', queryParams],
    queryFn: () => users.getAll(queryParams),
    enabled: Boolean(queryParams),
    onSuccess: (data) => {
      console.log('fetch-users-success', data)
    }
  })
```

Whenever i create/update/delete users, i use `refetch` method to get the latest data. However, `refetch` doesn't seem to be working for some reason. The api call is definitely made (can see the api call in network tab) but for some reason, the `fetchStatus` is always stuck in `fetching` after the `refetch()` method call. 

<img width="1113" alt="Screenshot 2023-09-17 at 3 19 39 PM" src="https://github.com/TanStack/query/assets/24563569/5c3aa47c-a589-4260-9d8e-0205cb60884c">

This only happens after refetch method call. What am i doing wrong?

**Comment by Lisba** [comment]
Hi @kcsujeet!

To get the fresh data after a mutation (create/update/delete), it is recommended to use [query invalidations](https://tanstack.com/query/v4/docs/react/guides/query-invalidation), this way you'd have a more declarative control over when it's needed to refetch data already stale because you just modified it.

The [query invalidations](https://tanstack.com/query/v4/docs/react/guides/query-invalidation) api supports partial query matching, so you can match multiple queries by their prefix, or get really specific and match an exact query with the `exact: true` option, but If you find yourself looking for even more granularity, you can use a predicate function where you could specify custom or more complex rules for invalidate queries.

On the other hand, keep in mind that the `onSuccess`, `onError` and `onSettled` callbacks are deprecated on the v4 and removed on v5 so avoid using them 😉

**Reply by kcsujeet** [reply]
Hi @Lisba , thank you for the reply. I did try that as well but the result was same. Api call was being made but `useQuery` was stuck in `fetching` state. 

Also, doesn't `refetch` invalidate the queries underneath, resulting in essentially the same behavior?

**Reply by Lisba** [reply]
Hi @kcsujeet, If the problem persists try debugging the fetching process itself (more than tanstack query), some possible approaches that occurs to me are:
1. First and foremost check your network tab into the chrome devTool to rule out a possible loop (or fetching data too often) causing the constant fetching state.
2. Set a low timeout to avoid network calls that take too much time.
3. Try using a simulated QueryFn, that is, instead of passing a function that fetches data through HTTP calls, pass a function that just returns a promise in order to check if the behaviour persists.
4. Try another TanStack Query version.

If none of the aforementioned actions seems to clarify the issue, try to reproduce the bug in a sandbox in order for me to check it and help you with more context.

**Reply by kcsujeet** [reply]
Hi @Lisba . I modified the code as below to show that the fetching process is actually fine. 

```
queryFn: async () => {
      const response = await userResource.getAll(queryParams)

      console.log('api-response', response.normalizedData)

      return response
}
```

here's the entire code:
````
export const useUsers = (view: Partial<View>): QueryResponse<User[]> => {
  const queryParams = useApiQueryParams({ view: view })

  const { data, isLoading, isError, refetch, fetchStatus, isRefetchError } = useQuery({
    queryKey: ['users'],
    queryFn: async () => {
      const response = await userResource.getAll(queryParams)

      console.log('api-response', response.normalizedData)

      return response
    }
  })

  console.log('fetch-users', data?.normalizedData, fetchStatus, isRefetchError)

  return {
    isLoading,
    isError,
    total: data?.meta?.records || 0,
    pages: data?.meta?.pages || 0,
    data: data?.normalizedData || [],
    refetch
  }
}
````

Here's the console log.

<img width="1113" alt="Screenshot 2023-09-18 at 10 06 45 AM" src="https://github.com/TanStack/query/assets/24563569/362758e9-f2e6-433a-9cbe-752f96aa2473">


As you can see in the screenshot above, the api response is received but for some reason `useQuery` doesn't resolve the `fetchStatus`. It get's stuck at `fetching`. 

I did try to replicate this problem in sandbox. But couldn't. Here's he sanbox [link](https://codesandbox.io/s/throbbing-bash-cxzhrt)

**Reply by TkDodo** [reply]
> queryKey: ['users'],

`queryParams` need to be part of the `queryKey`. You've shown this in the initial code snippet, but not in the last one ...

**Reply by kcsujeet** [reply]
Hi @TkDodo . 

i did try both. Both bring about same results. 

i also tried to invalidate the queries. redefined the refetch export as: 

`refetch: () => queryClient.invalidateQueries(['users', queryParams])`

this also didn't work.

**Reply by TkDodo** [reply]
since you have a working sandbox, I'd suggest you add code to that sandbox until the code looks like your codebase. You should see where it breaks along the way.

**Reply by kcsujeet** [reply]
[This](https://codesandbox.io/s/throbbing-bash-cxzhrt?file=/src/usePosts.ts) is the exact setup i have. Can't replicate it in the sandbox for some reason. 

under what conditions could the `fetchStatus` status be stuck on `fetching`?

**Reply by TkDodo** [reply]
> under what conditions could the fetchStatus status be stuck on fetching?

it shouldn't.

**Reply by kcsujeet** [reply]
update: 

it seems the refetch works if i don't pass `queryParams` to the `queryFn`. ie. if i just do

```
queryFn: () => users.getAll()
```
it seems to work. 

the `getAll` definition is as follows: 

```
public getAll = async (params?: ApiQueryParams): Promise<ApiResponse<T[]>> => {
    const res = await api.request<ApiResponse<T[]>>({
      method: HTTPMethod.GET,
      url: this.getPath(),
      params: params
    })

    return res.data
  }
```

**Reply by kcsujeet** [reply]
update: 

the culprit seems to be `include` property in `params`. 

```
  const params = {
    filter: {
      deactivated: {
        eq: false
      },
      organization_id: {
        eq: 3
      }
    },
    include: 'organization_memberships',
    sort: 'last_name'
  }
```

if i remove `include`. it works. 

@TkDodo do you think it has something to do with the query cache?

**Reply by kcsujeet** [reply]
if i rename `include` to `included`, it seems to work for some reason.

**Reply by kcsujeet** [reply]
update:

seems the issue lies somewhere in the data deserializer. We have an `axios` interceptor defined as: 

```
api.interceptors.response.use(
  function (response) {
    try {
      const dataFormatter = new Jsona()

      response.data.normalizedData = dataFormatter.deserialize(response.data)
    } catch (err) {
      console.error(err)
    }

    return response
  },
  function (error) {
    return Promise.reject(error)
  }
)

```

the `included` data is deserialized in this interceptor. could this be a race condition bug? ...

**Reply by kcsujeet** [reply]
@Lisba @TkDodo  i was finally able to replicate the issue. definitely something to do with the deserialization using `Jsona`. Here's the [link](https://codesandbox.io/s/throbbing-bash-cxzhrt?file=/api.ts) to sandbox.

<img width="1455" alt="Screenshot 2023-09-22 at 5 42 06 PM" src="https://github.com/TanStack/query/assets/24563569/da519f07-6539-4067-b7fe-c825418c2af0">

  
As you can see in the screenshot above, the `fetchStatus` is stuck in `fetching`.  I'd appreciate any help in this regards. 

Here's the `Jsona` documentation [link](https://github.com/olosegres/jsona#deserialize---creates-simplified-objects-from-json).

**Reply by TkDodo** [reply, marked as answer]
your normalized data circularly references itself via `organization`, `owner` and `user`. Our `structuralSharing` never finishes. Fix this or set `structuralSharing: false` (but you'll get more re-renders without it).

**Reply by kcsujeet** [reply]
@TkDodo ... does that mean react query is incompatible with data which is in [json-api](https://jsonapi.org/) format? because this could commonly happen in that format. an entity can have some other entity which in turn can belong to the previous entity ... any workarounds you would suggest?

**Reply by TkDodo** [reply]
> any workarounds

`structuralSharing: false`

**Reply by kcsujeet** [reply]
thank you @TkDodo  ... i appreciate all your responses. Really one of the most responsive maintainers in any lib I've used. 

I was able to solve the issue without disabling the `structuralSharing`. Just had to modify the behavior of the deserializer i'm using. For anyone interested, [here's](https://github.com/olosegres/jsona/issues/65#issuecomment-1751363481) what i did.
